### PR TITLE
[ROCm] No fence optimization to jit reduce template.

### PR DESCRIPTION
**Pull request by anatoliylitv**
Removing need for fences in normalization kernel by converting the stores into atomics+return. This is crucial for perf in architectures with split caches (e.g. MI300), where fences are inherently costly.

Reproducer:

```
import torch

shapes = [(4096, 8192)]


for shape in shapes:
    lo, hi = 0.999, 1.001
    x = lo + (hi - lo) * torch.randn(shape, device='cuda', dtype=torch.torch.float32)
    for _ in range(20):
        _ = torch.prod(x)
    torch.cuda.synchronize()

    start_evt = torch.cuda.Event(enable_timing=True)
    end_evt = torch.cuda.Event(enable_timing=True)
    start_evt.record()
    for _ in range(100):
        _ = torch.prod(x)
    end_evt.record()
    torch.cuda.synchronize()
    print(f"Avg time for shape {shape}: {start_evt.elapsed_time(end_evt) / 100 * 1e3:.2f} us")
```
**Before:**
Avg time for shape (4096, 8192): **112.59 us**
**After:**
Avg time for shape (4096, 8192): **41.05 us**

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang

### CODE REVIEW

File: aten/src/ATen/native/cuda/reduction_template.cuh [hunk 1]
```diff
@@ -4,6 +4,7 @@ const std::string reduction_template_0 = R"ESCAPE(
   #define C10_HOST_DEVICE __host__ __device__
   #define C10_DEVICE __device__
   #if defined(__clang__) && defined(__HIP__)
+  #define USE_ROCM 1
```
Comment (jeffdaily) [inline aten/src/ATen/native/cuda/reduction_template.cuh:7]: USE_ROCM should be defined by CMake during the build, not here.  Is this leftover from some debugging?

**Review by jeffdaily — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/cuda/reduction_template.cuh (hunk 1 above)
Comment (anatoliylitv) [reply aten/src/ATen/native/cuda/reduction_template.cuh:7]: @jeffdaily Unfortunately, this is not leftover code; it is a design issue in the current just‑in‑time (JIT) compilation implementation. What you are seeing is a single large string constant that contains a copy of code that also exists elsewhere, and is compiled at runtime.
All required definitions (for example, C10_HOST_DEVICE) are defined directly inside this string. In some parts of the string, USE_ROCM was replaced with `defined(__clang__) && defined(__HIP__)`, but not consistently across all instances.
A mechanism similar to hipify—which would automatically keep the JIT string in sync with the original source code and CMake definitions—has not yet been implemented. As a result, correctness currently depends on manual synchronization of changes.

```
const std::string reduction_template_0 = R"ESCAPE(
  #define C10_HOST_DEVICE __host__ __device__
  #define C10_DEVICE __device__
  #if defined(__clang__) && defined(__HIP__)
  #define USE_ROCM 1
```

**Review by jeffdaily — APPROVED** [review]
Approving — thanks for the explanation, @anatoliylitv.
